index.js

```diff
@@ -221,3 +221,61 @@ if (crypto) {
     activator
   );
 }
+
+if(global.Promise) {
+  var Promise = global.Promise;
+
+  global.Promise = function wrappedPromise(executor) {
+    if (!(this instanceof global.Promise)) {
+      return Promise(executor);
+    }
+
+    var promise = new Promise(wrappedExecutor);
+    var resolveWrapper;
+    var args;
+
+    executor.apply(this, args);
+
+    return promise;
+
+    function wrappedExecutor(accept, reject) {
+      args = [wrappedAccept, wrappedReject];
+
+      function wrappedAccept(val) {
+        if(promise.__asl_wrapper) return accept(val);
+        promise.__asl_wrapper = wrapCallback(function(fn, result) {
+          return fn.call(this, result);
+        });
+        return accept(val);
+      }
+
+      function wrappedReject(val) {
+        if(promise.__asl_wrapper) return reject(val);
+        promise.__asl_wrapper = wrapCallback(function(fn, result) {
+          return fn.call(this, result);
+        });
+        return reject(val);
+      }
+    }
+  }
+
+  var originalThen = Promise.prototype.then
+  Promise.prototype.then = wrappedThen
+
+  function wrappedThen() {
```

Comment:
I think since block scoping is on by default, having function declarations inside `if` is ok